File: packages/rag-pipeline/src/assistant_rh_rag_pipeline/context_builder.py
"""
Context builder for the RAG V3 Clean pipeline.

Selects sections to include in the LLM prompt under a fixed **token budget**,
applying:
  - Document-entire inclusion for small documents (< threshold)
  - Source triangulation (always include top sections from secondary publishers)
  - Legal-reference injection from ``rag_sections.references_juridiques``

The output is a list of ``ContextItem`` ready for prompt formatting.

Dependencies (internal only):
  - config (ContextBuildConfig, get_dsn)
  - models (AggregatedSection, ContextItem)
"""

from __future__ import annotations

import json
import logging
from collections import defaultdict
from typing import Dict, List, Optional, Set

import psycopg
from psycopg.rows import dict_row

from .config import ContextBuildConfig
from .db_helpers import get_dsn
from .models import AggregatedSection, ContextItem, estimate_tokens

logger = logging.getLogger(__name__)


class ContextBuilder:
    """
    Build the final context for the LLM generator from ranked sections.

    Strategy:
      1. If the top document is small (< ``doc_entire_threshold`` tokens),
         include **all its sections** up to ``max_full_docs``.
      2. Fill remaining budget with the highest-scored individual sections.
      3. **Triangulation**: always inject at least ``triangulation_sections``
         sections from publishers *other* than the top one.
      4. Append legal references cited inside the selected sections.
    """

    def __init__(self, config: ContextBuildConfig, dsn: str | None = None):
        self.config = config
        self.dsn = dsn or get_dsn()
        self.last_resolved_refs: Dict[str, Dict[str, str]] = {}

    def build(self, sections: List[AggregatedSection]) -> List[ContextItem]:
        if not sections:
            return []

        budget = self.config.get_token_budget()
        max_full_docs = self.config.get_max_full_docs()
        doc_threshold = self.config.get_doc_entire_threshold()
        max_sections = self.config.get_max_sections()
        refs_budget = self.config.get_legal_refs_budget()

        selected: List[ContextItem] = []
        used_ids: Set[str] = set()
        tokens_used = 0
        full_doc_count = 0

        # Group by document (only sections that have a document_id). Sections
        # without a document_id are picked up by the Step-2 loop directly.
        by_doc: Dict[str, List[AggregatedSection]] = defaultdict(list)
        for s in sections:
            if s.document_id:
                by_doc[str(s.document_id)].append(s)

        # Sort doc groups by best section score (descending); compute the max
        # once and reuse it (used both for sort and for the inner gate).
        doc_max_score: Dict[str, float] = {doc_id: max((s.score or 0.0) for s in secs) for doc_id, secs in by_doc.items()}
        sorted_docs = sorted(by_doc.items(), key=lambda kv: doc_max_score[kv[0]], reverse=True)

        def _section_key(s: AggregatedSection) -> str:
            # Unique-and-stable across both loops. `id(s)` is fine because the
            # input list is stable for the lifetime of this build().
            return s.section_id or s.heading or f"_obj_{id(s)}"

        # Step 1 – doc-entire for small documents. We always include qualifying
        # small docs (within budget and the max_full_docs cap); the original
        # legifrance-q2 ordering bug is fixed by the score-sort at the end of
        # Step 2 below, not by suppressing doc-entires here.
        for doc_id, doc_sections in sorted_docs:
            if full_doc_count >= max_full_docs:
                break
            doc_token_count = doc_sections[0].metadata.get("doc_token_count", 0) or 0
            if doc_token_count <= 0 or doc_token_count > doc_threshold:
                continue
            if tokens_used + doc_token_count > budget:
                continue

            doc_row = self._load_full_document(doc_id)
            if not doc_row or not doc_row.get("doc_markdown"):
                continue

            item = self._full_doc_to_item(doc_row, doc_sections)
            item.metadata["is_doc_entire"] = True
            # ContextItem.score on a doc-entire reflects the best section score
            # of its constituents, so the post-Step-2 sort places it correctly.
            item.score = doc_max_score[doc_id]
            selected.append(item)
            for s in doc_sections:
                used_ids.add(_section_key(s))
            tokens_used += item.token_estimate
            full_doc_count += 1
            logger.info("Doc-entire included doc %s (%s): ~%d tokens", doc_id, item.document_title[:40], item.token_estimate)

        # Step 2 – fill with top individual sections (from selector/reranker order)
        for s in sections:
            key = _section_key(s)
            if key in used_ids:
                continue
            item = self._section_to_item(s)
            if tokens_used + item.token_estimate > budget:
                continue
            if len(selected) >= max_sections:
                break
            selected.append(item)
            used_ids.add(key)
            tokens_used += item.token_estimate

        # The sort key puts highest-scoring items first; on exact score ties,
        # standalone chunks beat doc-entires (promotion intent from review
        # finding #7) and Python's stable sort preserves retrieval order for
        # the remaining ties.
        def _sort_key(item):
            return (-(item.score or 0.0), 1 if item.metadata.get("is_doc_entire") else 0)

        # Determine the "primary publisher" for triangulation BEFORE appending —
        # use the highest-scored section in the input rather than `selected[0]`
        # so triangulation depends on the data, not on insertion ordering.
        primary_publisher = None
        if sections:
            top_section = max(sections, key=lambda s: s.score or 0.0)
            primary_publisher = top_section.publisher

        # Step 3 – triangulation (ignores budget to guarantee publisher diversity)
        tri_added = 0
        for s in sections:
            if tri_added >= self.config.triangulation_sections:
                break
            if s.publisher == primary_publisher:
                continue
            key = _section_key(s)
            if key in used_ids:
                continue
            item = self._section_to_item(s)
            item.metadata["is_triangulation"] = True
            selected.append(item)
            used_ids.add(key)
            tokens_used += item.token_estimate
            tri_added += 1

        # Single sort at the very end so triangulation items participate in the
        # final ordering — the LLM always sees the strongest evidence first,
        # regardless of which step contributed each item.
        selected.sort(key=_sort_key)

        # Step 4 – resolve legal references from rag_chunks_dgafp and inject
        refs_tokens = 0
        all_ref_numbers = self._collect_ref_numbers(selected)
        cid_map = self._resolve_cids(all_ref_numbers) if all_ref_numbers else {}
        self.last_resolved_refs = cid_map

        # Allocate the legal-refs budget primary-content-first: triangulation
        # items are publisher-diversity fillers and must not consume refs_budget
        # ahead of the answer-bearing primary sources, even when they outscore
        # them. Stable sort preserves the score order within each group, so the
        # prompt order (``selected``) is untouched — only the refs pass reorders.
        refs_order = sorted(selected, key=lambda it: bool(it.metadata.get("is_triangulation")))
        for item in refs_order:
            if item.references_juridiques and refs_tokens < refs_budget:
                self._enrich_refs_with_cid(item, cid_map)
                ref_text = self._format_references(item.references_juridiques)
                ref_tokens = estimate_tokens(ref_text)
                if refs_tokens + ref_tokens <= refs_budget:
                    item.content += f"\n\n---\nReferences juridiques :\n{ref_text}"
                    item.token_estimate += ref_tokens
                    tokens_used += ref_tokens
                    refs_tokens += ref_tokens

        logger.info(
            "Context built: %d items, ~%d tokens (budget %d, mode %s), %d full docs, %d legal refs",
            len(selected),
            tokens_used,
            budget,
            self.config.context_mode.value,
            full_doc_count,
            len(cid_map),
        )
        return selected

    # ------------------------------------------------------------------
    # Formatting helpers
    # ------------------------------------------------------------------

    @staticmethod
    def format_for_prompt(items: List[ContextItem]) -> str:
        """Render context items as a string ready for the LLM user-message."""
        parts: List[str] = []
        for i, item in enumerate(items, 1):
            header = f"[Source {i}] {item.document_title or item.heading}"
            if item.publisher:
                header += f" ({item.publisher})"
            parts.append(f"### {header}\n\n```markdown\n{item.content}\n```\n\n---\n")
        return "\n".join(parts)

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _load_full_document(self, doc_id: str) -> Optional[dict]:
        """Load the full document markdown and metadata from rag_documents."""
        sql = """
            SELECT doc_id, title, source_url, publisher, doc_markdown, token_count
            FROM rag_documents
            WHERE doc_id = %s AND doc_markdown IS NOT NULL
        """
        try:
            with psycopg.connect(self.dsn, row_factory=dict_row) as conn:
                return conn.execute(sql, (doc_id,)).fetchone()
        except psycopg.Error as exc:
            logger.warning("Failed to load full document %s: %s", doc_id, exc)
            return None

    def _full_doc_to_item(self, doc_row: dict, matched_sections: List[AggregatedSection]) -> ContextItem:
        """Convert a rag_documents row into a single ContextItem with the full document content."""
        best_score = max((s.score for s in matched_sections), default=0.0)
        all_refs = []
        for s in matched_sections:
            if s.references_juridiques:
                all_refs.append(s.references_juridiques)
        return ContextItem(
            section_id=None,
            heading=doc_row.get("title", ""),
            content=doc_row.get("doc_markdown", ""),
            score=best_score,
            publisher=doc_row.get("publisher", ""),
            document_title=doc_row.get("title", ""),
            document_url=doc_row.get("source_url"),
            references_juridiques=all_refs[0] if all_refs else None,
            token_estimate=doc_row.get("token_count") or estimate_tokens(doc_row.get("doc_markdown", "")),
            metadata={
                "doc_id": str(doc_row.get("doc_id", "")),
                "doc_title": doc_row.get("title", ""),
                "doc_url": doc_row.get("source_url"),
                "doc_publisher": doc_row.get("publisher", ""),
                "doc_token_count": doc_row.get("token_count", 0),
            },
        )

    def _section_to_item(self, s: AggregatedSection) -> ContextItem:
        return ContextItem(
            section_id=s.section_id,
            heading=s.heading,
            content=s.markdown,
            score=s.score,
            publisher=s.publisher,
            document_title=s.metadata.get("doc_title", ""),
            document_url=s.metadata.get("doc_url"),
            references_juridiques=s.references_juridiques,
            token_estimate=s.token_estimate,
            metadata=s.metadata,
        )

    @staticmethod
    def _collect_ref_numbers(items: List[ContextItem]) -> List[str]:
        """Extract all article numbers from references_juridiques across items."""
        numbers = []
        for item in items:
            refs = item.references_juridiques
            if not refs:
                continue
            if isinstance(refs, str):
                try:
                    refs = json.loads(refs)
                except (ValueError, json.JSONDecodeError):
                    continue
            if isinstance(refs, list):
                for r in refs:
                    if isinstance(r, dict) and r.get("number"):
                        numbers.append(r["number"])
        return list(set(numbers))

    def _resolve_cids(self, numbers: List[str]) -> Dict[str, Dict[str, str]]:
        """Lookup article numbers in rag_chunks_dgafp to get cid and url."""
        if not numbers:
            return {}
        sql = """
            SELECT number, cid, url, full_title
            FROM rag_chunks_dgafp
            WHERE number = ANY(%s)
            GROUP BY number, cid, url, full_title
        """
        try:
            with psycopg.connect(self.dsn, row_factory=dict_row) as conn:
                rows = conn.execute(sql, (numbers,)).fetchall()
            return {r["number"]: {"cid": r["cid"] or "", "url": r["url"] or "", "title": r["full_title"] or ""} for r in rows if r.get("cid")}
        except psycopg.Error as exc:
            logger.warning("CID resolution failed: %s", exc)
            return {}

    @staticmethod
    def _enrich_refs_with_cid(item: ContextItem, cid_map: Dict[str, Dict[str, str]]):
        """Add cid/url to references_juridiques entries if found in cid_map."""
        refs = item.references_juridiques
        if not refs:
            return
        if isinstance(refs, str):
            try:
                refs = json.loads(refs)
            except (ValueError, json.JSONDecodeError):
                return
        if isinstance(refs, list):
            for r in refs:
                if isinstance(r, dict) and r.get("number") in cid_map:
                    r.update(cid_map[r["number"]])
            item.references_juridiques = refs

    @staticmethod
    def _format_references(refs) -> str:
        if isinstance(refs, str):
            return refs
        if isinstance(refs, dict):
            lines = []
            for key, val in refs.items():
                lines.append(f"- {key}: {val}")
            return "\n".join(lines)
        if isinstance(refs, list):
            return "\n".join(f"- {r}" for r in refs)
        return str(refs)

File: apps/mastra-pipeline/src/mastra/steps/context-builder.ts
import { createStep } from "@mastra/core/workflows";
import { z } from "zod";
import {
	DEFAULT_RUNTIME_RAG_CONFIG,
	getRuntimeRagConfig,
	type RuntimeRagConfig,
} from "../lib/config";
import { getDbPool } from "../lib/db";
import {
	contextBuildConfigSchema,
	contextSelectorStateSchema,
	contextSelectorStepInputSchema,
} from "./context-selector";
import { retrieverStepInputSchema } from "./retriever";
import { aggregatedSectionSchema, sectionAggregatorStepInputSchema } from "./section-aggregator";

interface NormalizedContextBuildConfig {
	context_mode: "standard" | "wide";
	token_budget: number;
	max_full_docs: number;
	doc_entire_threshold: number;
	max_sections: number;
	triangulation_sections: number;
	legal_refs_budget: number;
	token_budget_wide: number;
	max_full_docs_wide: number;
	doc_entire_threshold_wide: number;
	max_sections_wide: number;
	legal_refs_budget_wide: number;
}

interface FullDocumentRow {
	doc_id: string;
	title: string | null;
	source_url: string | null;
	publisher: string | null;
	doc_markdown: string | null;
	token_count: number | null;
}

interface LegalReferenceRow {
	number: string | null;
	cid: string | null;
	url: string | null;
	full_title: string | null;
}

const contextItemSchema = z.object({
	sectionId: z.string().nullable(),
	heading: z.string(),
	content: z.string(),
	score: z.number(),
	publisher: z.string().nullable(),
	documentTitle: z.string().nullable(),
	documentUrl: z.string().nullable(),
	referencesJuridiques: z.unknown().nullable(),
	tokenEstimate: z.number().int().nonnegative(),
	metadata: z.record(z.string(), z.unknown()),
});

const contextMetaSchema = z.object({
	contextMode: z.enum(["standard", "wide"]),
	tokenBudget: z.number().int().positive(),
	tokenCount: z.number().int().nonnegative(),
	refsTokenCount: z.number().int().nonnegative(),
	maxSections: z.number().int().positive(),
	selectedCount: z.number().int().nonnegative(),
	fullDocCount: z.number().int().nonnegative(),
	triangulationAdded: z.number().int().nonnegative(),
	legalRefsResolvedCount: z.number().int().nonnegative(),
	legalRefsInjectedCount: z.number().int().nonnegative(),
	lastResolvedRefs: z.record(z.string(), z.record(z.string(), z.string())),
	warnings: z.array(z.string()),
});

export const contextBuilderStepInputSchema = z.object({
	sections: z.array(aggregatedSectionSchema),
	config: contextBuildConfigSchema.optional(),
});

export const contextBuilderStepOutputSchema = z.object({
	contextItems: z.array(contextItemSchema),
	context: z.string(),
	contextMeta: contextMetaSchema,
});

export const contextBuilderStateSchema = contextSelectorStateSchema.extend({
	config: z
		.object({
			retrieval: retrieverStepInputSchema.shape.config.optional(),
			aggregation: sectionAggregatorStepInputSchema.shape.config.optional(),
			selector: contextSelectorStepInputSchema.shape.config.optional(),
			context: contextBuildConfigSchema.optional(),
		})
		.passthrough()
		.optional(),
	contextBuilder: contextBuilderStepOutputSchema.optional(),
});

type ContextBuilderInput = z.infer<typeof contextBuilderStepInputSchema>;
type ContextBuilderOutput = z.infer<typeof contextBuilderStepOutputSchema>;
type ContextBuildConfigInput = z.infer<typeof contextBuildConfigSchema>;
type ContextItem = z.infer<typeof contextItemSchema>;
type AggregatedSection = z.infer<typeof aggregatedSectionSchema>;

function estimateTokens(text: string): number {
	return Math.floor((text ?? "").length / 4);
}

function normalizeContextBuildConfig(
	runtimeConfig: RuntimeRagConfig["context"],
	override?: ContextBuildConfigInput,
): NormalizedContextBuildConfig {
	const base: NormalizedContextBuildConfig = {
		context_mode: "standard",
		token_budget: 8000,
		max_full_docs: 1,
		doc_entire_threshold: 3500,
		max_sections: 12,
		triangulation_sections: 2,
		legal_refs_budget: 1000,
		token_budget_wide: 12000,
		max_full_docs_wide: 2,
		doc_entire_threshold_wide: 5000,
		max_sections_wide: 20,
		legal_refs_budget_wide: 2000,
	};

	return {
		context_mode: override?.context_mode ?? runtimeConfig?.context_mode ?? base.context_mode,
		token_budget: override?.token_budget ?? runtimeConfig?.token_budget ?? base.token_budget,
		max_full_docs: override?.max_full_docs ?? runtimeConfig?.max_full_docs ?? base.max_full_docs,
		doc_entire_threshold:
			override?.doc_entire_threshold ??
			runtimeConfig?.doc_entire_threshold ??
			base.doc_entire_threshold,
		max_sections: override?.max_sections ?? runtimeConfig?.max_sections ?? base.max_sections,
		triangulation_sections:
			override?.triangulation_sections ??
			runtimeConfig?.triangulation_sections ??
			base.triangulation_sections,
		legal_refs_budget:
			override?.legal_refs_budget ?? runtimeConfig?.legal_refs_budget ?? base.legal_refs_budget,
		token_budget_wide:
			override?.token_budget_wide ?? runtimeConfig?.token_budget_wide ?? base.token_budget_wide,
		max_full_docs_wide:
			override?.max_full_docs_wide ?? runtimeConfig?.max_full_docs_wide ?? base.max_full_docs_wide,
		doc_entire_threshold_wide:
			override?.doc_entire_threshold_wide ??
			runtimeConfig?.doc_entire_threshold_wide ??
			base.doc_entire_threshold_wide,
		max_sections_wide:
			override?.max_sections_wide ?? runtimeConfig?.max_sections_wide ?? base.max_sections_wide,
		legal_refs_budget_wide:
			override?.legal_refs_budget_wide ??
			runtimeConfig?.legal_refs_budget_wide ??
			base.legal_refs_budget_wide,
	};
}

async function loadRuntimeContextBuildConfigSafe(): Promise<RuntimeRagConfig["context"]> {
	try {
		const runtimeConfig = await getRuntimeRagConfig();
		return runtimeConfig.context;
	} catch {
		return DEFAULT_RUNTIME_RAG_CONFIG.context;
	}
}

function resolveContextModeValues(config: NormalizedContextBuildConfig): {
	mode: "standard" | "wide";
	tokenBudget: number;
	maxFullDocs: number;
	docEntireThreshold: number;
	maxSections: number;
	legalRefsBudget: number;
} {
	if (config.context_mode === "wide") {
		return {
			mode: "wide",
			tokenBudget: config.token_budget_wide,
			maxFullDocs: config.max_full_docs_wide,
			docEntireThreshold: config.doc_entire_threshold_wide,
			maxSections: config.max_sections_wide,
			legalRefsBudget: config.legal_refs_budget_wide,
		};
	}

	return {
		mode: "standard",
		tokenBudget: config.token_budget,
		maxFullDocs: config.max_full_docs,
		docEntireThreshold: config.doc_entire_threshold,
		maxSections: config.max_sections,
		legalRefsBudget: config.legal_refs_budget,
	};
}

async function loadFullDocument(docId: string): Promise<FullDocumentRow | null> {
	const db = getDbPool();
	const result = await db.query<FullDocumentRow>(
		`
      SELECT
        doc_id::text AS doc_id,
        title,
        source_url,
        publisher,
        doc_markdown,
        token_count
      FROM rag_documents
      WHERE doc_id = $1::uuid
        AND doc_markdown IS NOT NULL
      LIMIT 1
    `,
		[docId],
	);

	if (result.rowCount === 0) {
		return null;
	}

	return result.rows[0] ?? null;
}

function asNumber(value: unknown, fallback = 0): number {
	if (typeof value === "number" && Number.isFinite(value)) {
		return value;
	}

	const parsed = Number.parseFloat(String(value));
	return Number.isFinite(parsed) ? parsed : fallback;
}

function asNullableString(value: unknown): string | null {
	if (typeof value !== "string") {
		return null;
	}

	const normalized = value.trim();
	return normalized.length > 0 ? normalized : null;
}

function toSectionContextItem(section: AggregatedSection): ContextItem {
	return {
		sectionId: section.sectionId,
		heading: section.heading,
		content: section.markdown,
		score: section.score,
		publisher: section.publisher,
		documentTitle: asNullableString(section.metadata.doc_title),
		documentUrl: asNullableString(section.metadata.doc_url),
		referencesJuridiques: section.referencesJuridiques,
		tokenEstimate: estimateTokens(section.markdown),
		metadata: section.metadata,
	};
}

function toFullDocContextItem(
	docRow: FullDocumentRow,
	matchedSections: AggregatedSection[],
): ContextItem {
	const bestScore = Math.max(...matchedSections.map((section) => section.score));

	const allReferences = matchedSections
		.map((section) => section.referencesJuridiques)
		.filter((refs) => refs !== null && refs !== undefined);

	const markdown = docRow.doc_markdown ?? "";
	return {
		sectionId: null,
		heading: docRow.title ?? "",
		content: markdown,
		score: bestScore,
		publisher: docRow.publisher,
		documentTitle: docRow.title,
		documentUrl: docRow.source_url,
		referencesJuridiques: allReferences[0] ?? null,
		tokenEstimate: docRow.token_count ?? estimateTokens(markdown),
		metadata: {
			doc_id: docRow.doc_id,
			doc_title: docRow.title ?? "",
			doc_url: docRow.source_url ?? "",
			doc_publisher: docRow.publisher ?? "",
			doc_token_count: docRow.token_count ?? 0,
		},
	};
}

function collectReferenceNumbers(items: ContextItem[]): string[] {
	const numbers = new Set<string>();

	for (const item of items) {
		const refs = item.referencesJuridiques;
		if (!refs) {
			continue;
		}

		let parsed: unknown = refs;
		if (typeof refs === "string") {
			try {
				parsed = JSON.parse(refs);
			} catch {
				continue;
			}
		}

		if (!Array.isArray(parsed)) {
			continue;
		}

		for (const entry of parsed) {
			if (!entry || typeof entry !== "object" || Array.isArray(entry)) {
				continue;
			}

			const number = asNullableString((entry as Record<string, unknown>).number);
			if (number) {
				numbers.add(number);
			}
		}
	}

	return Array.from(numbers);
}

async function resolveLegalReferenceCids(
	numbers: string[],
): Promise<Record<string, { cid: string; url: string; title: string }>> {
	if (numbers.length === 0) {
		return {};
	}

	const db = getDbPool();
	const result = await db.query<LegalReferenceRow>(
		`
      SELECT
        number,
        cid,
        url,
        full_title
      FROM rag_chunks_dgafp
      WHERE number = ANY($1::text[])
      GROUP BY number, cid, url, full_title
    `,
		[numbers],
	);

	const out: Record<string, { cid: string; url: string; title: string }> = {};
	for (const row of result.rows) {
		const number = asNullableString(row.number);
		const cid = asNullableString(row.cid);

		if (!number || !cid) {
			continue;
		}

		out[number] = {
			cid,
			url: asNullableString(row.url) ?? "",
			title: asNullableString(row.full_title) ?? "",
		};
	}

	return out;
}

function enrichReferencesWithCid(
	refs: unknown,
	cidMap: Record<string, { cid: string; url: string; title: string }>,
): unknown {
	let parsed: unknown = refs;

	if (typeof parsed === "string") {
		try {
			parsed = JSON.parse(parsed);
		} catch {
			return refs;
		}
	}

	if (!Array.isArray(parsed)) {
		return refs;
	}

	const enriched = parsed.map((entry) => {
		if (!entry || typeof entry !== "object" || Array.isArray(entry)) {
			return entry;
		}

		const item = { ...(entry as Record<string, unknown>) };
		const number = asNullableString(item.number);

		if (number && cidMap[number]) {
			item.cid = cidMap[number].cid;
			item.url = cidMap[number].url;
			item.title = cidMap[number].title;
		}

		return item;
	});

	return enriched;
}

function formatReferences(refs: unknown): string {
	if (typeof refs === "string") {
		return refs;
	}

	if (refs && typeof refs === "object" && !Array.isArray(refs)) {
		return Object.entries(refs)
			.map(([key, value]) => `- ${key}: ${String(value)}`)
			.join("\n");
	}

	if (Array.isArray(refs)) {
		return refs.map((entry) => `- ${JSON.stringify(entry)}`).join("\n");
	}

	return String(refs ?? "");
}

export function formatContextForPrompt(items: ContextItem[]): string {
	return items
		.map((item, idx) => {
			let header = `[Source ${idx + 1}] ${item.documentTitle || item.heading}`;
			if (item.publisher) {
				header += ` (${item.publisher})`;
			}

			return `### ${header}\n\n\`\`\`markdown\n${item.content}\n\`\`\`\n\n---\n`;
		})
		.join("\n");
}

export async function runContextBuilder(
	input: ContextBuilderInput,
	runtimeContextConfigOverride?: RuntimeRagConfig["context"],
): Promise<ContextBuilderOutput> {
	const runtimeContextConfig =
		runtimeContextConfigOverride ?? (await loadRuntimeContextBuildConfigSafe());
	const config = normalizeContextBuildConfig(runtimeContextConfig, input.config);
	const modeValues = resolveContextModeValues(config);

	if (input.sections.length === 0) {
		return {
			contextItems: [],
			context: "",
			contextMeta: {
				contextMode: modeValues.mode,
				tokenBudget: modeValues.tokenBudget,
				tokenCount: 0,
				refsTokenCount: 0,
				maxSections: modeValues.maxSections,
				selectedCount: 0,
				fullDocCount: 0,
				triangulationAdded: 0,
				legalRefsResolvedCount: 0,
				legalRefsInjectedCount: 0,
				lastResolvedRefs: {},
				warnings: [],
			},
		};
	}

	const selected: ContextItem[] = [];
	const usedIds = new Set<string>();
	let tokensUsed = 0;
	let fullDocCount = 0;
	let triangulationAdded = 0;
	let refsTokenCount = 0;
	let legalRefsInjectedCount = 0;
	const warnings: string[] = [];

	const byDoc = new Map<string, AggregatedSection[]>();
	const standalone: AggregatedSection[] = [];

	for (const section of input.sections) {
		if (section.documentId) {
			const existing = byDoc.get(section.documentId);
			if (existing) {
				existing.push(section);
			} else {
				byDoc.set(section.documentId, [section]);
			}
			continue;
		}

		standalone.push(section);
	}

	const sortedDocs = Array.from(byDoc.entries()).sort((left, right) => {
		const leftScore = Math.max(...left[1].map((section) => section.score));
		const rightScore = Math.max(...right[1].map((section) => section.score));
		return rightScore - leftScore;
	});

	for (const [docId, docSections] of sortedDocs) {
		if (fullDocCount >= modeValues.maxFullDocs) {
			break;
		}

		const docTokenCount = asNumber(docSections[0]?.metadata.doc_token_count, 0);
		if (docTokenCount <= 0 || docTokenCount > modeValues.docEntireThreshold) {
			continue;
		}

		if (tokensUsed + docTokenCount > modeValues.tokenBudget) {
			continue;
		}

		try {
			const docRow = await loadFullDocument(docId);
			if (!docRow?.doc_markdown) {
				continue;
			}

			const fullDocItem = toFullDocContextItem(docRow, docSections);
			fullDocItem.metadata.is_doc_entire = true;

			selected.push(fullDocItem);
			for (const section of docSections) {
				usedIds.add(section.sectionId || section.heading);
			}

			tokensUsed += fullDocItem.tokenEstimate;
			fullDocCount += 1;
		} catch (error) {
			const message = error instanceof Error ? error.message : "Unknown full-doc load error";
			warnings.push(`Failed to load full doc ${docId}: ${message}`);
		}
	}

	for (const section of input.sections) {
		const key = section.sectionId || section.heading;
		if (usedIds.has(key)) {
			continue;
		}

		const item = toSectionContextItem(section);
		if (tokensUsed + item.tokenEstimate > modeValues.tokenBudget) {
			continue;
		}

		if (selected.length >= modeValues.maxSections) {
			break;
		}

		selected.push(item);
		usedIds.add(key);
		tokensUsed += item.tokenEstimate;
	}

	for (const section of standalone) {
		const key = section.sectionId || section.heading || String(selected.length);
		if (usedIds.has(key)) {
			continue;
		}

		const item = toSectionContextItem(section);
		if (tokensUsed + item.tokenEstimate > modeValues.tokenBudget) {
			continue;
		}

		if (selected.length >= modeValues.maxSections) {
			break;
		}

		selected.push(item);
		usedIds.add(key);
		tokensUsed += item.tokenEstimate;
	}

	const primaryPublisher = selected[0]?.publisher;
	for (const section of input.sections) {
		if (triangulationAdded >= config.triangulation_sections) {
			break;
		}

		if (section.publisher === primaryPublisher) {
			continue;
		}

		const key = section.sectionId || section.heading;
		if (usedIds.has(key)) {
			continue;
		}

		const item = toSectionContextItem(section);
		if (tokensUsed + item.tokenEstimate > modeValues.tokenBudget) {
			continue;
		}

		selected.push(item);
		usedIds.add(key);
		tokensUsed += item.tokenEstimate;
		triangulationAdded += 1;
	}

	const allRefNumbers = collectReferenceNumbers(selected);

	let cidMap: Record<string, { cid: string; url: string; title: string }> = {};
	try {
		cidMap = await resolveLegalReferenceCids(allRefNumbers);
	} catch (error) {
		const message = error instanceof Error ? error.message : "Unknown CID resolution error";
		warnings.push(`CID resolution failed: ${message}`);
	}

	for (const item of selected) {
		if (!item.referencesJuridiques || refsTokenCount >= modeValues.legalRefsBudget) {
			continue;
		}

		item.referencesJuridiques = enrichReferencesWithCid(item.referencesJuridiques, cidMap);
		const refText = formatReferences(item.referencesJuridiques);
		const refTokens = estimateTokens(refText);

		if (
			refsTokenCount + refTokens > modeValues.legalRefsBudget ||
			tokensUsed + refTokens > modeValues.tokenBudget
		) {
			continue;
		}

		item.content += `\n\n---\nReferences juridiques :\n${refText}`;
		item.tokenEstimate += refTokens;
		refsTokenCount += refTokens;
		tokensUsed += refTokens;
		legalRefsInjectedCount += 1;
	}

	return {
		contextItems: selected,
		context: formatContextForPrompt(selected),
		contextMeta: {
			contextMode: modeValues.mode,
			tokenBudget: modeValues.tokenBudget,
			tokenCount: tokensUsed,
			refsTokenCount,
			maxSections: modeValues.maxSections,
			selectedCount: selected.length,
			fullDocCount,
			triangulationAdded,
			legalRefsResolvedCount: Object.keys(cidMap).length,
			legalRefsInjectedCount,
			lastResolvedRefs: cidMap,
			warnings,
		},
	};
}

export const contextBuilderStep = createStep({
	id: "context-builder",
	description: "Build final context items and prompt payload from selected sections",
	inputSchema: contextBuilderStepInputSchema,
	outputSchema: contextBuilderStepOutputSchema,
	stateSchema: contextBuilderStateSchema,
	execute: async ({ inputData, state, setState }) => {
		const runtimeContextConfig = await loadRuntimeContextBuildConfigSafe();
		const result = await runContextBuilder(inputData, runtimeContextConfig);

		await setState({
			...state,
			contextBuilder: result,
		});

		return result;
	},
});
